src/cart/crt_context.c

```diff
@@ -1516,8 +1521,12 @@ crt_context_req_untrack(struct crt_rpc_priv *rpc_priv)
 			epi->epi_req_num++;
 			D_ASSERT(epi->epi_req_num >= epi->epi_reply_num);
 
-			/* add to resend list */
-			d_list_add_tail(&tmp_rpc->crp_tmp_link, &submit_list);
+			/* add to submit list if not cancelled or timed out already  */
+			if (tmp_rpc->crp_timeout_ts != 0) {
+				/* prevent rpc from being released before it is dispatched below */
+				RPC_ADDREF(tmp_rpc);
+				d_list_add_tail(&tmp_rpc->crp_tmp_link_submit, &submit_list);
```

Comment:
[Not request for changes] I suggest we assert `d_list_empty(&tmp_rpc->crp_tmp_link_submit)`.